internal/stoppropaganda/websites.go

```diff
@@ -400,114 +432,108 @@ func startWebsites() {
+
+}
+func (website *Website) ValidateDNS() {
+	// Lock, because we don't want multiple validations
+	// running at the same time
+	website.validateMux.Lock()
+	defer website.validateMux.Unlock()
+
+	if time.Since(website.dnsLastChecked) >= VALIDATE_DNS_EVERY {
+		ipAddresses, err := customresolver.GetIPs(website.host)
+		website.dnsLastChecked = time.Now()
+
+		if err != nil {
+			errStr := err.Error()
+			if strings.HasSuffix(errStr, "Temporary failure in name resolution") || strings.HasSuffix(errStr, "connection refused") {
+				website.SchedulePause(1*time.Second, "Your DNS servers unreachable or returned an error: "+errStr)
+				return
+			}
+
+			reason := errStr
+			switch {
+			case strings.HasSuffix(err.Error(), "Temporary failure in name resolution"):
+				reason = "Your DNS servers unreachable or returned an error"
+			case strings.HasSuffix(err.Error(), "connection refused"):
+				reason = "Your DNS servers reachable, but refusing connections"
+			case strings.HasSuffix(errStr, "no such host"):
+				reason = "Domain does not exist: " + errStr
+			case strings.HasSuffix(errStr, "No address associated with hostname"):
+				reason = "Domain does not have any IPs assigned: " + errStr
+			}
+
+			website.SchedulePause(10*time.Second, reason)
+			return
+		}
+
+		nonpublicerr := resolvefix.CheckNonPublicIP(ipAddresses)
+		if nonpublicerr != nil {
+			website.SchedulePause(5*time.Minute, nonpublicerr.Error())
+			return
+		}
+	}
+}
+
 func runWebsiteWorker(c chan *Website) {
 	// Each worker has it's own response
 	req := fasthttp.AcquireRequest()
 	resp := fasthttp.AcquireResponse()
 
 	for {
 		website := <-c
-		website.req.CopyTo(req) // https://github.com/valyala/fasthttp/issues/53#issuecomment-185125823
-
-		website.pauseMux.Lock()
-		if time.Since(website.dnsLastChecked) >= VALIDATE_DNS_EVERY {
-			ipAddresses, err := getIPs(website.host)
-			website.dnsLastChecked = time.Now()
-
-			if err != nil {
-				website.mux.Lock()
-				switch {
-				case strings.HasSuffix(err.Error(), "Temporary failure in name resolution"):
-					website.Status = "Your DNS servers unreachable or returned an error"
-				case strings.HasSuffix(err.Error(), "connection refused"):
-					website.Status = "Your DNS servers reachable, but refusing connections"
-				case strings.HasSuffix(err.Error(), "no such host"):
-					website.Status = "Domain does not exist"
-				case strings.HasSuffix(err.Error(), "No address associated with hostname"):
-					website.Status = "Domain does not have any IPs assigned"
-				default:
-					website.Status = err.Error()
-				}
-				website.mux.Unlock()
-
-				time.Sleep(10 * time.Second)
-				website.pauseMux.Unlock()
-				continue
-			}
-
-			if containsNonPublicIP(ipAddresses) {
-				website.mux.Lock()
-				website.Status = "Non public IP detected"
-				website.paused = true
-				website.mux.Unlock()
-
-				time.Sleep(5 * time.Minute)
-				website.pauseMux.Unlock()
-				continue
-			}
-
+		if !website.ShouldRun() {
+			continue
```

Comment:
Why don't we sleep for e.g. 1 second here? Without sleeping even for 100ms we are literally wasting CPU.